### PR TITLE
DM-42576: mpSkyEphemerisQuery

### PR DESCRIPTION
**Pull request by Gerenjie**
Adding mpSkyEphemerisQuery, which replaces SkyBot in querying mpSky for ssObjects in the pre-load step.

### CODE REVIEW

**Comment by timj** [comment]
@Gerenjie the reason the action is failing is because the DM repos require rebasing and not merging from main in order to keep the ticket branch up to date. If you need help with that you can Slack me.

**Review by kfindeisen — CHANGES_REQUESTED** [review]
Thanks for taking the time to put this together! I have quite a few comments, so I'm returning the review ahead of the other PRs. My most important objections are the timeout and the `print` statements; leaving either alone will make this task almost impossible to use in the field. Please also make sure you're [building the docs](https://developer.lsst.io/stack/building-single-package-docs.html); I saw some formatting problems that I would expect to cause a build error.

Happy to look at this again once these comments are addressed.

File: python/lsst/ap/ephemerides.yaml
[the remark this answers is not in the document]
Comment (kfindeisen) [reply python/lsst/ap/ephemerides.yaml:1]: I see that lsst/ap_pipe#189 has its own `Ephemerides.yaml`, so I recommend just dropping this file entirely.

File: python/lsst/ap/association/mpSkyEphemerisQuery.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+        doc="IAU Minor Planet Center observer code for queries "
+            "(Rubin Obs./LSST default is X05)",
+        default='X05'
+    )
+    queryRadiusDegrees = pexConfig.Field(
+        dtype=float,
+        doc="On sky radius for Ephemeris cone search. Defaults "
+            "to the radius of Rubin Obs FoV in degrees",
+        default=1.75
+    )
```
[the remark this answers is not in the document]
Comment (kfindeisen) [reply python/lsst/ap/association/mpSkyEphemerisQuery.py:84]: Turns out there's no good documentation about making config overrides. 😞 A quick from-memory summary:

* The files in `obs_*/config/` contain config settings that should be applied for all runs of that task on that instrument. They have higher priority than task defaults, but lower priority than in-pipeline overrides.
* The config file must have the same name as the task label in the pipeline (which is usually, including in your case, `_DefaultName`).
* Overrides directly under `obs_lsst/config/` apply to _all_ LSST instruments. Overrides in a subdirectory apply only to that instrument, and have higher priority than the general overrides.

[the remark this answers is not in the document]
Comment (kfindeisen) [reply python/lsst/ap/association/mpSkyEphemerisQuery.py:84]: On second glance, something is very wrong here -- you don't actually _use_ `observerCode` when uploading to MPSky. Is this a future feature?

File: python/lsst/ap/association/mpSkyEphemerisQuery.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+            if nFound == 0:
+                self.log.info("No Solar System objects found for visit.")
+
+            self.log.info("%d Solar System Objects in visit", nFound)
+        except requests.exceptions.ConnectionError as e:
+            print("failed to connect to the remote ephemerides service. details:", file=sys.stderr)
+            print(e, file=sys.stderr)
+            MPSkySsObjects = pd.DataFrame(
+                columns=['ObjID', 'ra', 'dec', 'obj_poly', 'obs_poly',
+                         'Err(arcsec)', 'ssObjectId'])
```
[the remark this answers is not in the document]
Comment (kfindeisen) [reply python/lsst/ap/association/mpSkyEphemerisQuery.py:216]: DM-45457 is now done, so you can make the solar system table optional if that's what you were going for.

File: python/lsst/ap/association/mpSkyEphemerisQuery.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Solar System Object Query to MPSky in place of a internal Rubin solar
+system object caching/retrieval code.
+
+Will compute the location for of known SSObjects within a visit. This code
+blocks on web requests, so should not be used as part of any real-time or
+time-sensitive system. Use in a larger pipeline at your own risk.
+"""
```
[the remark this answers is not in the document]
Comment (Gerenjie) [reply python/lsst/ap/association/mpSkyEphemerisQuery.py:28]: These lines were inherited from SkyBotEphemerisQuery, which I used as a template for this task, and inadvertently left in. The warning is no longer true -- we're changing from SkyBot to MPSky specifically to solve this problem.

File: python/lsst/ap/association/mpSkyEphemerisQuery.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+            dec = r["dec"].to_numpy()
+
+            MPSkySsObjects = pd.DataFrame()
+            MPSkySsObjects['ObjID'] = ObjID
+            MPSkySsObjects['ra'] = ra
+            MPSkySsObjects['dec'] = dec
+            MPSkySsObjects['obj_poly'] = list(np.zeros((len(MPSkySsObjects), 5)))  # fix, eventually
+            MPSkySsObjects['obs_poly'] = list(np.zeros((len(MPSkySsObjects), 5)))  # fix, eventually
+            MPSkySsObjects['Err(arcsec)'] = 2
+            MPSkySsObjects['ssObjectId'] = [abs(hash(v)) for v in MPSkySsObjects['ObjID'].values]
```
[the remark this answers is not in the document]
Comment (Gerenjie) [reply python/lsst/ap/association/mpSkyEphemerisQuery.py:204]: If we discover twice as many SSObjects as expected and end up with 10 million rows, then the odds of nonzero collisions are (10 million)^2 / (2^63), which is around 1 in 100,000. Not sure what our risk tolerance is here -- 1 in 100,000 is very small but definitely non-zero.

[the remark this answers is not in the document]
Comment (kfindeisen) [reply python/lsst/ap/association/mpSkyEphemerisQuery.py:204]: This doesn't need to be solved right away -- it's easy enough to change -- but there was a lot of care put into how source and object IDs are generated to make sure they're always unique, have support for future features like versioned catalogs, etc. ([RFC-785](https://rubinobs.atlassian.net/browse/RFC-785) is the best documentation I can find, though it's more about code organization than the actual algorithms). The solar system group might want to have a similar plan in place before real observations start and the IDs are locked in.

File: python/lsst/ap/association/ssoAssociation.py [hunk 5]
```diff
@@ -121,18 +123,18 @@ def run(self, diaSourceCatalog, solarSystemObjects, exposure):
...
             # Which DIA Sources fall within r?
             dist, idx = tree.query(ssoVect, distance_upper_bound=maxRadius)
             if np.isfinite(dist[0]):
                 nFound += 1
-                diaSourceCatalog.loc[idx[0], "ssObjectId"] = ssObject["ssObjectId"]
+                diaSourceCatalog.loc[diaSourceCatalog.index[idx[0]], "ssObjectId"] = ssObject["ssObjectId"]
 
         self.log.info("Successfully associated %d SolarSystemObjects.", nFound)
-        assocMask = diaSourceCatalog["ssObjectId"] != 0
+        assocMask = (diaSourceCatalog["ssObjectId"] != 0) & (np.isfinite(diaSourceCatalog["ssObjectId"]))
```
[the remark this answers is not in the document]
Comment (Gerenjie) [reply python/lsst/ap/association/ssoAssociation.py:137]: I don't know if ssObjectID has defined behavior for diaSources which are not associated to an ssObject.

[the remark this answers is not in the document]
Comment (Gerenjie) [reply python/lsst/ap/association/ssoAssociation.py:137]: Disregard that -- I do define it and set it to zero. I agree, this change is not needed.

[the remark this answers is not in the document]
Comment (Gerenjie) [reply python/lsst/ap/association/ssoAssociation.py:137]: Disregard that -- I do define it and set it to zero. I agree, this change is not needed.

File: python/lsst/ap/association/mpSkyEphemerisQuery.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Gerenjie) [reply python/lsst/ap/association/mpSkyEphemerisQuery.py:84]: We should use X05 for AuxTel for now. AuxTel getting an observatory code is blocked on association working, since the Minor Planet Center assigns observatory codes and they've asked for some NEO measurements. I don't see an obs_auxtel equivalent though -- should that be somewhere in obs_lsst?

Now that I look at it, I don't see a SkyBotEphemerisQuery.py override in obs_lsst at all. Should I follow the format from obs_subaru and obs_decam?

[the remark this answers is not in the document]
Comment (Gerenjie) [reply python/lsst/ap/association/mpSkyEphemerisQuery.py:84]: Yes, observerCode is a future MPSky feature, but it currently just defaults to LSST (X05).

File: python/lsst/ap/association/mpSkyEphemerisQuery.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        dtype=str,
+        doc="IAU Minor Planet Center observer code for queries "
+            "(Rubin Obs./LSST default is X05)",
+        default='X05'
+    )
+    queryRadiusDegrees = pexConfig.Field(
+        dtype=float,
+        doc="On sky radius for Ephemeris cone search. Defaults "
+            "to the radius of Rubin Obs FoV in degrees",
+        default=1.75
```
[the remark this answers is not in the document]
Comment (Gerenjie) [reply python/lsst/ap/association/mpSkyEphemerisQuery.py:83]: The padding value is a non-obvious choice that we might want to change in the future, so I think it should be a config. Not sure what the default should be, but happy to reduce it a bit.

File: python/lsst/ap/association/mpSkyEphemerisQuery.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+                ``obs_poly``
+                    DO NOT USE until t_min issue is resolved
+        """
+        # Grab the visitInfo from the raw to get the information needed on the
+        # full visit.
+
+        # Midpoint time of the exposure in JD
+        region = predictedRegionTime.region
+        timespan = predictedRegionTime.timespan
+        expCenter = SpherePoint(region.getBoundingCircle().getCenter())
```
[the remark this answers is not in the document]
Comment (Gerenjie) [reply python/lsst/ap/association/mpSkyEphemerisQuery.py:141]: We could change the config parameter to 'buffer radius' and use the bounding circle radius here, or leave as-is. I believe the two are functionally identical and I don't feel strongly one way or the other -- do you have a preference?

File: python/lsst/ap/association/mpSkyEphemerisQuery.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kfindeisen) [reply python/lsst/ap/association/mpSkyEphemerisQuery.py:84]: > I don't see an obs_auxtel equivalent though -- should that be somewhere in obs_lsst?

Yes, that's what the subdirectories are for. I _think_ `LATISS` is the only one that's for AuxTel instead of Simonyi?

> Now that I look at it, I don't see a SkyBotEphemerisQuery.py override in obs_lsst at all. Should I follow the format from obs_subaru and obs_decam?

The [default for SkyBotEphemerisQuery was I11](https://github.com/lsst/ap_association/blob/main/python/lsst/ap/association/skyBotEphemerisQuery.py#L80), which was already for LSST. I'm not sure how I11 relates to X05.

Where you _do_ need an override (LATISS only?), then yes, following the existing format should be enough.

File: python/lsst/ap/association/mpSkyEphemerisQuery.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (kfindeisen) [reply python/lsst/ap/association/mpSkyEphemerisQuery.py:141]: I'd prefer to make use of the actual camera geometry as much as possible -- it's less error-prone (no need to look up and configure instrument-specific numbers!) and lets us use much tighter tolerances.

In fact, I suspect the size of the buffer will be set by the objects' proper motion (given a few minutes' uncertainty in future observation time), and not by the telescope's pointing accuracy, FOV, or other specs.

File: python/lsst/ap/association/mpSkyEphemerisQuery.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kfindeisen) [reply python/lsst/ap/association/mpSkyEphemerisQuery.py:84]: > Yes, observerCode is a future MPSky feature, but it currently just defaults to LSST (X05).

Please add a note to the field docstring that it's not used yet.

This does raise the question of what to do for DECam and especially HSC. We could just run the task and accept that nothing will match (that's what `SkyBotEphemerisQueryTask` did before I learned that geocentric is _not_ a good default 😛), or we could trim the task out of those pipelines until MPSky supports them.

File: python/lsst/ap/association/mpSkyEphemerisQuery.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+
+        # Make sure date is non-NaN.
+        expMidPointEPOCH = (timespan.begin.mjd + timespan.end.mjd)/2
```
Comment (kfindeisen) [inline python/lsst/ap/association/mpSkyEphemerisQuery.py:144]: After reviewing #230, I don't think this will work -- `astropy.time.Time` doesn't support the `+` operator. Fortunately, once that PR merges, you should be able to just do:
```suggestion
        expMidPointEPOCH = utils.getMidpointFromTimespan(timespan).mjd
```
As a bonus, `getMidpointFromTimespan` would do the input validation for you.

File: python/lsst/ap/association/mpSkyEphemerisQuery.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Gerenjie) [reply python/lsst/ap/association/mpSkyEphemerisQuery.py:84]: I11 is Gemini South, which is basically LSST as far as positions on the Earth are concerned. X05 probably didn't exist as a code when this was written (it was just designated last year, I believe), but as far as science is concerned these are all equivalent except in very extreme cases (impactors). 

I've created DM-45735, and will cite it in comments! Should be resolved soon, but who knows.

File: python/lsst/ap/association/mpSkyEphemerisQuery.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Gerenjie) [reply python/lsst/ap/association/mpSkyEphemerisQuery.py:202]: After reading DM-45457, I'm still not sure exactly how to make this change. Could we meet sometime on Friday to discuss?

**Review by kfindeisen — CHANGES_REQUESTED** [review]
Looks much better, most of my comments are about documentation. However, I'm a bit concerned about the exception handling and the correctness of the ID converter; can you add some round-trip tests for the latter?

File: python/lsst/ap/association/diaPipe.py [hunk 9]
```diff
@@ -67,9 +67,10 @@ class DiaPipelineConnections(
     solarSystemObjectTable = connTypes.Input(
         doc="Catalog of SolarSolarSystem objects expected to be observable in "
             "this detectorVisit.",
-        name="visitSsObjects",
+        name="preloaded_SsObjects",
```
[the remark this answers is not in the document]
Comment (Gerenjie) [reply python/lsst/ap/association/diaPipe.py:70]: Yes -- Ian recommended that based on his and Meredith's discussion, `preloaded_` would be the most consistent.

File: python/lsst/ap/association/diaPipe.py [hunk 10]
```diff
@@ -405,8 +408,8 @@ def run(self,
         ----------
         diaSourceTable : `pandas.DataFrame`
             Newly detected DiaSources.
-        solarSystemObjectTable : `pandas.DataFrame`
-            Preloaded Solar System objects expected to be visible in the image.
+        legacySolarSystemTable : `pandas.DataFrame`
+            Not used
```
[the remark this answers is not in the document]
Comment (Gerenjie) [reply python/lsst/ap/association/diaPipe.py:412]: The idea is to maintain the interface's backwards compatibility -- removing the argument in the place of `legacySolarSystemTable` would break every place where `diaPipe.run` is currently being called. I do not believe there is a plan for phasing this out.

File: python/lsst/ap/association/mpSkyEphemerisQuery.py [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+        dtype=str,
+        doc="IAU Minor Planet Center observer code for queries "
+            "(Rubin Obs./LSST default is X05)",
+        default='X05'
+    )
+    queryBufferRadiusDegrees = pexConfig.Field(
+        dtype=float,
+        doc="Buffer radius added to camera bounding circle for Ephemeris "
+            "cone search. Defaults to 10 deg/day * 30 minutes",
+        default=0.208
```
[the remark this answers is not in the document]
Comment (Gerenjie) [reply python/lsst/ap/association/mpSkyEphemerisQuery.py:77]: Yes. It might be overkill, but I don't know of a safe upper bound for the difference between a predictedRegionTime and the actual observation it produces. If this causes performance issues I'm happy to reduce it.

File: python/lsst/ap/association/mpSkyEphemerisQuery.py [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+            response = requests.get(mpSkyURL, params=params, timeout=self.config.mpSkyRequestTimeoutSeconds)
+            response.raise_for_status()
+            with pa.input_stream(memoryview(response.content)) as fp:
+                fp.seek(0)
+                object_polynomial = pa.ipc.read_tensor(fp)
+                observer_polynomial = pa.ipc.read_tensor(fp)
+                with pa.ipc.open_stream(fp) as reader:
+                    row = next(reader)
+
+            ObjID = row["name"].to_numpy(zero_copy_only=False)
```
[the remark this answers is not in the document]
Comment (Gerenjie) [reply python/lsst/ap/association/mpSkyEphemerisQuery.py:180]: @mjuric -- I'll need your help renaming this slice of code.

File: python/lsst/ap/association/diaPipe.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (kfindeisen) [reply python/lsst/ap/association/diaPipe.py:412]: Except that what you have here is not actually backward-compatible -- a hypothetical user who tries to call `run` with the old signature, expecting to get solar system associations, will silently get nothing. If you _do_ want backward compatibility, you'd need to convert the old-style input into the new-style input.

Phasing out could be handled [by raising a `FutureWarning`](https://developer.lsst.io/stack/deprecating-interfaces.html) if the argument isn't `None` (there's no specific procedure for deprecating parameters, but see that page for recommended text) and filing a Jira issue to remove it later.

All that said, though, I doubt that `run` is called anywhere except from `runQuantum` or unit tests, so I think you could get away with unceremoniously deleting it.

File: python/lsst/ap/association/utils.py [hunk 13]
```diff
@@ -261,3 +261,71 @@ def getMidpointFromTimespan(timespan, allowUnbounded=True):
...
+    if len(ObjID) > 7:
+        raise ValueError(f'ObjID longer than 7 characters: "{ObjID}"')
+    if flags < 0 or flags > 255:
+        raise ValueError(f'Flags ({flags}) outside [0, 255].')
+
+    ssObjectID = flags
+    for character in ObjID:
+        ssObjectID *= 256
+        ssObjectID += ord(character)
+    ssObjectID *= 256 ** (7 - len(ObjID))
```
[the remark this answers is not in the document]
Comment (Gerenjie) [reply python/lsst/ap/association/utils.py:301]: That's the hope! Yes, the idea is for everyone to use these methods to convert, and for RSP users to have to use table lookups to check the ssObjectID for the minor planet they care about.

File: python/lsst/ap/association/mpSkyEphemerisQuery.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (kfindeisen) [reply python/lsst/ap/association/mpSkyEphemerisQuery.py:77]: Fine, let's wait and see. Grabbing a roughly raft-size area is still a big improvement over the whole focal plane.

File: python/lsst/ap/association/diaPipe.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (kfindeisen) [reply python/lsst/ap/association/diaPipe.py:412]: After discussing with @Gerenjie and @isullivan separately, we've agreed to move forward as follows:

- `legacySolarSystemTable` will be used in place of `solarSystemObjectTable` when the latter is undefined. That should be enough to give backwards-compatibility whether `run` is called with positional args or keywords, and is the only change that will be made right now.
- _After_ merging this PR, we'll file an RFC to deprecate and remove `legacySolarSystemTable` [per the usual process](https://developer.lsst.io/stack/deprecating-interfaces.html). I doubt there will be any pushback.

File: tests/test_utils.py [hunk 14]
```diff
@@ -94,3 +94,44 @@ def test_invalidTimespan(self):
...
+        allowed_strings = [ch for ch in 'ABCDEFGHIJKMNOPQRSTUVWXYZ0123456789 -'] \
+            + ['A' * i for i in range(2, 8)] + ['Z' * i for i in range(2, 8)] \
+            + ['J95X00A', 'J95X01L', 'J95F13B', 'J98SA8Q', 'J98SC7V', 'J98SG2S'] \
+            + ['K99AJ3Z', 'K08Aa0A', 'K07Tf8A', 'PLS2040', 'T1S3138', 'T2S1010', 'T3S4101']
+        allowed_flags = [i for i in range(0, 256)]
+        for allowed_string in allowed_strings:
+            for allowed_flag in allowed_flags:
+                self.assertEqual((allowed_string, allowed_flag),
+                                 ssObjectID_to_objID(objID_to_ssObjectID(allowed_string,
+                                                                         allowed_flag)))
```
Comment (kfindeisen) [inline tests/test_utils.py:110]: Might be worth factoring the `ssObjectID_to_objID(objID_to_ssObjectID())` call into its own variable for readability.

File: tests/test_utils.py [hunk 15]
```diff
@@ -94,3 +94,44 @@ def test_invalidTimespan(self):
...
+    def test_invalid_ssObjectID_to_objID_and_objID_to_ssObjectID(self):
+        """Convert between ssObjectIDs and MPC packed designations
+        """
+        allowed_strings = [ch for ch in 'ABCDEFGHIJKMNOPQRSTUVWXYZ0123456789 -'] \
+                + ['A' * i for i in range(2, 8)] + ['Z' * i for i in range(2, 8)] \
+                + ['J95X00A', 'J95X01L', 'J95F13B', 'J98SA8Q', 'J98SC7V', 'J98SG2S'] \
+                + ['K99AJ3Z', 'K08Aa0A', 'K07Tf8A', 'PLS2040', 'T1S3138', 'T2S1010', 'T3S4101']
+        allowed_flags = [i for i in range(0, 256)]
+        disallowed_flags = [-999999999, -512, -256, -255, -1, 256, 512, 99999999]
+        disallowed_strings = ['Ā', '🔭', 'A' * 8, ' ' * 8, 'A' * 128]
```
Comment (kfindeisen) [inline tests/test_utils.py:121]: I think putting an emoji in Python code might be going a little too far (though it's hilarious that GitHub can print it!)

Comment (kfindeisen) [reply tests/test_utils.py:121]: What about the empty string? That seems like a natural edge case.

File: tests/test_utils.py [hunk 16]
```diff
@@ -94,3 +94,44 @@ def test_invalidTimespan(self):
             getMidpointFromTimespan(timespan_none_both, allowUnbounded=True)
         with self.assertRaises(ValueError):
             getMidpointFromTimespan(timespan_none_both, allowUnbounded=False)
+
+    def test_ssObjectID_to_objID_and_objID_to_ssObjectID(self):
+        """Convert between ssObjectIDs and MPC packed designations
+        """
+        allowed_strings = [ch for ch in 'ABCDEFGHIJKMNOPQRSTUVWXYZ0123456789 -'] \
+            + ['A' * i for i in range(2, 8)] + ['Z' * i for i in range(2, 8)] \
```
Comment (kfindeisen) [inline tests/test_utils.py:102]: Why are single-character strings not allowed?

File: tests/test_utils.py [hunk 17]
```diff
@@ -113,9 +113,9 @@ def test_invalid_ssObjectID_to_objID_and_objID_to_ssObjectID(self):
         """Convert between ssObjectIDs and MPC packed designations
         """
         allowed_strings = [ch for ch in 'ABCDEFGHIJKMNOPQRSTUVWXYZ0123456789 -'] \
-                + ['A' * i for i in range(2, 8)] + ['Z' * i for i in range(2, 8)] \
-                + ['J95X00A', 'J95X01L', 'J95F13B', 'J98SA8Q', 'J98SC7V', 'J98SG2S'] \
-                + ['K99AJ3Z', 'K08Aa0A', 'K07Tf8A', 'PLS2040', 'T1S3138', 'T2S1010', 'T3S4101']
+            + ['A' * i for i in range(2, 8)] + ['Z' * i for i in range(2, 8)] \
+            + ['J95X00A', 'J95X01L', 'J95F13B', 'J98SA8Q', 'J98SC7V', 'J98SG2S'] \
+            + ['K99AJ3Z', 'K08Aa0A', 'K07Tf8A', 'PLS2040', 'T1S3138', 'T2S1010', 'T3S4101']
```
Comment (kfindeisen) [inline tests/test_utils.py:118]: I assume this breaks `flake8`, so this commit should have been part of "adding unit tests"

**Review by kfindeisen — APPROVED** [review]
Looks good, only minor comments.

However, please squash the commits before merging:
- "response to review" commits [are still discouraged](https://developer.lsst.io/work/flow.html#resolving-a-review)
- unit tests should be added on the same commit as the code they test, otherwise the commits are not self-contained (imagine if somebody deleted or rearranged one of the commits -- you'd have either untested code, or tests of nonexistent code). That said, the ID converters _would_ have made a good separate commit _a priori_, since they can be defined and tested without MPSkyEphemerisQuery.
- fixes needed to get things to build should be merged into the original change, as if it had been done right the first time; as things stand, the original test commit breaks `scons`/`flake8`.

"Made legacySolarSystemTable an optional input" is a bit of a grey area; this one makes sense either as a standalone commit or squashed under the "right the first time" principle. I'd lean towards squashing (as it's basically a bug in the original `legacySolarSystemTable`), but if you keep it separate, please [write it in the imperative tense](https://developer.lsst.io/work/flow.html#writing-commit-summary-lines), e.g. "Use legacySolarSystemTable in new code".

File: tests/test_utils.py (hunk 16 above)
Comment (Gerenjie) [reply tests/test_utils.py:102]: They are allowed.

Comment (kfindeisen) [reply tests/test_utils.py:102]: Oops, silly me. The first line. 🤦

File: python/lsst/ap/association/mpSkyEphemerisQuery.py [hunk 18]
```diff
@@ -167,25 +199,15 @@ def _mpSkyConeSearch(self, expCenter, epochMJD, queryRadius):
...
-            dec = r["dec"].to_numpy()
+            objID, ra, dec, object_polynomial, observer_polynomial = self.read_mpSky_response(response)
 
             mpSkySsObjects = pd.DataFrame()
-            mpSkySsObjects['ObjID'] = ObjID
+            mpSkySsObjects['ObjID'] = objID
             mpSkySsObjects['ra'] = ra
-            # TODO: accept MPSky polynomials when they are well-formed.
-            mpSkySsObjects['obs_poly'] = list(np.zeros((len(mpSkySsObjects), 5)))
-            mpSkySsObjects['obj_poly'] = list(np.zeros((len(mpSkySsObjects), 5)))
```
[the remark this answers is not in the document]
Comment (Gerenjie) [reply python/lsst/ap/association/mpSkyEphemerisQuery.py]: No downstream behavior depends on these values, and my previous choice of stand-in zeros was to make sure there was some data of the right type there. Anything downstream trying to interpret those zeros as actual values would have had major issues.

File: python/lsst/ap/association/utils.py [hunk 19]
```diff
@@ -313,19 +314,18 @@ def ssObjectID_to_ObjID(ssObjectID):
...
 
     Raises
     ------
-
     """
-    ObjID = ''.join([chr(ssObjectID // (256 ** i) % 256) for i in reversed(range(0, 7))])
-    ObjID = ObjID.replace('\x00', '')
-    return ObjID, ssObjectID//(256 ** 7)%256
+    if ssObjectID < 0 or ssObjectID >= (2 << 63):
+        raise ValueError(f'ssObjectID ({ssObjectID}) outside [0, 2^63].')
```
[the remark this answers is not in the document]
Comment (Gerenjie) [reply python/lsst/ap/association/utils.py:327]: Good catch!

**Comment by kfindeisen** [comment]
Just for future reference, the commit summaries overflowed GitHub's 72-character limit. See [the dev guide](https://developer.lsst.io/work/flow.html#appendix-commit-message-best-practices) for recommendations on writing commit messages (which actually recommends the much tighter limit of 50 characters, maybe we should update that).
